### PR TITLE
functions implementation and code refractor.

### PR DESCRIPTION
**Pull request by huojiao2006**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,23 +1,26 @@
...
 
 ENV GO111MODULE=on
 ENV CGO_ENABLED=0
 ENV GOOS=linux
 
 RUN mkdir -p /openpitrix_bin
-RUN go build -v -a -installsuffix cgo -ldflags '-w' -o /openpitrix_bin/notification-manager cmd/notification/main.go
+#to fix a issue: verifying ...: checksum mismatch
+RUN rm go.sum;go mod download
+RUN go build -v -a -installsuffix cgo -ldflags '-w' -o /openpitrix_bin/notification cmd/notification/main.go
```
Comment (chilianyi) [inline Dockerfile:19]: If add version, need `go generate openpitrix.io/notification/pkg/version` before build.

File: api/types.proto [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+import "google/protobuf/wrappers.proto";
+import "google/protobuf/timestamp.proto";
+
+message ErrorDetail {
+	string error_name = 1;
+	string cause = 2;
+}
+
+message ResourceCategory {
```
Comment (chilianyi) [inline api/types.proto]: Useless

File: deploy/db/mysql.yaml [hunk 3]
```diff
@@ -0,0 +1,89 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: notification-db
+  namespace: openpitrix-system
```
Comment (chilianyi) [inline deploy/db/mysql.yaml]: Other new system, such as notification-system.

File: deploy/db/mysql.yaml [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  labels:
+    app: notification
+    tier: db
+    version: test-1.0.0-alpha
```
Comment (chilianyi) [inline deploy/db/mysql.yaml]: Use Semantic Versioning: https://semver.org/

File: deploy/openpitrix-notification.yaml [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        image: busybox:1.28.4
+        imagePullPolicy: IfNotPresent
+        command: ['sh', '-c', 'until nc -z notification-db.openpitrix-system.svc 3306; do echo "waiting for mysql"; sleep 2; done;']
+      - name: wait-etcd
+        image: busybox:1.28.4
+        imagePullPolicy: IfNotPresent
+        command: ['sh', '-c', 'until nc -z notification-etcd.openpitrix-system.svc 2379; do echo "waiting for etcd"; sleep 2; done;']
+      containers:
+      - name: notification
+        image: huojiao2006/public_repodev:v0.0.1
```
Comment (chilianyi) [inline deploy/openpitrix-notification.yaml]: Use image in repo openpitrix

File: pkg/db/schema/notification/V0_2__init.sql [hunk 6]
```diff
@@ -0,0 +1,4 @@
+
+drop table nf_address_list;
```
Comment (chilianyi) [inline pkg/db/schema/notification/V0_2__init.sql:2]: Why drop the table? Change name?

File: pkg/gerr/error.go [hunk 7]
```diff
@@ -12,8 +12,8 @@ import (
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
 
-	"openpitrix.io/openpitrix/pkg/logger"
-	"openpitrix.io/openpitrix/pkg/pb"
+	"openpitrix.io/notification/pkg/logger"
```
Comment (chilianyi) [inline pkg/gerr/error.go:15]: Do not use "openpitrix.io/notification/pkg/logger" again.

File: pkg/logger/logger.go [hunk 8]
```diff
@@ -0,0 +1,199 @@
+// Copyright 2018 The OpenPitrix Authors. All rights reserved.
```
Comment (chilianyi) [inline pkg/logger/logger.go:1]: Useless.

File: pkg/models/address.go [hunk 9]
```diff
@@ -26,6 +26,19 @@ type Address struct {
 	Status           string    `gorm:"column:status"`
 }
 
+type AddressWithListID struct {
```
Comment (chilianyi) [inline pkg/models/address.go]: Rename to AddressWithListId

File: pkg/models/address.go [hunk 10]
```diff
@@ -93,3 +105,27 @@ func ParseAddressSet2PbSet(inAddrs []*Address) []*pb.Address {
 	}
 	return pbAddrs
 }
+
+func ParseAddressWithListIDSet2PbSet(inAddrs []*AddressWithListID) []*pb.Address {
```
Comment (chilianyi) [inline pkg/models/address.go]: Change ID to Id, unify the fmt.

File: pkg/services/notification/am.go [hunk 11]
```diff
@@ -7,18 +7,94 @@ package notification
 import (
 	"context"
 
+	"openpitrix.io/notification/pkg/models"
+
```
Comment (chilianyi) [inline pkg/services/notification/am.go]: Useless empty line.

File: pkg/services/notification/am.go [hunk 12]
```diff
@@ -7,18 +7,94 @@ package notification
...
 func (s *Server) Checker(ctx context.Context, req interface{}) error {
 	switch r := req.(type) {
 	case *pb.ServiceConfig:
 		return manager.NewChecker(ctx, r).
 			Required(constants.ServiceCfgProtocol, constants.ServiceCfgEmailHost, constants.ServiceCfgPort, constants.ServiceCfgDisplayEmail, constants.ServiceCfgEmail, constants.ServiceCfgPassword).
 			Exec()
+	case *pb.GetServiceConfigRequest:
+		return manager.NewChecker(ctx, r).
+			StringChosen("service_type", NfTypes).
```
Comment (chilianyi) [inline pkg/services/notification/am.go]: You can change all of the parameters into `constants.xxx`

File: pkg/services/notification/handler.go [hunk 13]
```diff
@@ -75,83 +79,187 @@ func (s *Server) CreateNotification(ctx context.Context, req *pb.CreateNotificat
 	if err != nil {
 		return nil, err
 	}
+	logger.Debugf(ctx, "Create Notification[%s] in DB successfully.", notification.NotificationId)
 
-	tasks, err := SplitNotificationIntoTasks(notification)
+	_, err = s.createTasksByNotification(ctx, notification)
+	if err != nil {
+		logger.Errorf(ctx, "create Tasks By Notification failed, [%+v].", err)
```
Comment (chilianyi) [inline pkg/services/notification/handler.go]: Change the log like this: Create tasks by notification failed.

File: pkg/services/notification/handler.go [hunk 14]
```diff
@@ -75,83 +79,187 @@ func (s *Server) CreateNotification(ctx context.Context, req *pb.CreateNotificat
...
-		NotificationId: pbutil.ToProtoString(notification.NotificationId),
-	}, nil
+	return tasks, nil
 }
 
 func (s *Server) DescribeNotifications(ctx context.Context, req *pb.DescribeNotificationsRequest) (*pb.DescribeNotificationsResponse, error) {
 	nfs, nfCnt, err := rs.DescribeNotifications(ctx, req)
 	if err != nil {
-		logger.Errorf(ctx, "Failed to describe notifications, error: %+v", err)
+		logger.Errorf(ctx, "Failed to Describe Notifications, [%+v], [%+v].", req, err)
```
Comment (chilianyi) [inline pkg/services/notification/handler.go]: No need to log req, the req will be log in grpc_server.

File: pkg/services/notification/handler.go [hunk 15]
```diff
@@ -75,83 +79,187 @@ func (s *Server) CreateNotification(ctx context.Context, req *pb.CreateNotificat
...
+	taskIds, err := rs.UpdateTasks2Pending(ctx, taskIds)
+	if err != nil {
+		logger.Errorf(ctx, "Failed to Retry Tasks[%+v], [%+v].", taskIds, err)
+		return nil, gerr.NewWithDetail(ctx, gerr.Internal, err, gerr.ErrorRetryTaskFailed, taskIds)
+	}
+
+	var taskReq = &pb.DescribeTasksRequest{
+		TaskId: taskIds,
+	}
+	tasks, _, err := rs.DescribeTasks(ctx, taskReq)
```
Comment (chilianyi) [inline pkg/services/notification/handler.go:253]: First describe, then update.

File: pkg/util/pbutil/pb.go [hunk 16]
```diff
@@ -7,12 +7,13 @@ package pbutil
 import (
 	"time"
 
+	"openpitrix.io/notification/pkg/constants"
```
Comment (chilianyi) [inline pkg/util/pbutil/pb.go]: fmt

**Comment by chilianyi** [comment]
有几个问题：
1. 命名规则尽量统一，比如列名尽量都用Column 开头。 用变量就都用变量。
2. logger 的格式，要是一个句子。 首字母大写，其他字母非特殊名称小写。 
3. req， struct这些不需要在log里面打印。 
4. 点开头的几个文件，不能删除， .travis 删了，所以才没办法构建， .env 是docker-compose.yaml 用到的。

File: deploy/openpitrix-notification.yaml (hunk 5 above)
Comment (huojiao2006) [reply deploy/openpitrix-notification.yaml]: updated.

File: pkg/db/schema/notification/V0_2__init.sql (hunk 6 above)
Comment (huojiao2006) [reply pkg/db/schema/notification/V0_2__init.sql:2]: will add it back in next update.

File: pkg/gerr/error.go (hunk 7 above)
Comment (huojiao2006) [reply pkg/gerr/error.go:15]: some functions not implemated in "openpitrix.io/logger", will upgraded "openpitrix.io/logger" in next update.

File: pkg/logger/logger.go (hunk 8 above)
Comment (huojiao2006) [reply pkg/logger/logger.go:1]: update.

File: pkg/models/address.go (hunk 9 above)
Comment (huojiao2006) [reply pkg/models/address.go]: remove it,  no need it.

File: pkg/models/address.go (hunk 10 above)
Comment (huojiao2006) [reply pkg/models/address.go]: remove it,  no need it.

File: pkg/services/notification/am.go (hunk 11 above)
Comment (huojiao2006) [reply pkg/services/notification/am.go]: updated

File: pkg/services/notification/am.go (hunk 12 above)
Comment (huojiao2006) [reply pkg/services/notification/am.go]: updated.

File: pkg/services/notification/handler.go (hunk 13 above)
Comment (huojiao2006) [reply pkg/services/notification/handler.go]: updated.

File: pkg/services/notification/handler.go (hunk 14 above)
Comment (huojiao2006) [reply pkg/services/notification/handler.go]: updated.

File: pkg/services/notification/handler.go (hunk 15 above)
Comment (huojiao2006) [reply pkg/services/notification/handler.go:253]: updated.

File: pkg/util/pbutil/pb.go (hunk 16 above)
Comment (huojiao2006) [reply pkg/util/pbutil/pb.go]: updated.

**Pull request merged**